### PR TITLE
NOREF - Fix race condition (and manual jobs)

### PR DESCRIPTION
**Pull request by nonword**
Fix race condition that may drop records when kinesis write time exceeds Sierra fetch time:

When more than `request_batch_size` (e.g. 100) results are received, the poller must fetch a second set of records. Writing the first set of records to kinesis occurs in parallel to fetching that second set of records. If Sierra takes N time to return the second set of records, but it takes N + 1 time to write the first set of records to Kinesis, the app (previously) nullified the second result set (held in @previous_results), making it impossible for the app to subsequently send the second result set through kinesis. Essentially, the app assumed that fetching records from Sierra will always be slower than writing records to Kinesis, which is not always true. This has caused the last page of records to sometimes be dropped when there are multiple pages (and kinesis writes are slow).

The fix is to recast our @previous_records (holding the most recently fetched records) as an array of resultsets, so that it can grow to hold multiple (although I suspect it will only ever have max 1 as written) buffered resultsets if Kinesis is slow to write. In the scenario above, instead of nullifying @previous_results after writing the records to Kinesis, the kinesis-writing code shifts the resultset off of @previous_results right before beginning the kinesis-write, allowing the Sierra-retreiving code to safely push additional resultsets onto @previous_results without risk that they'll be erroneously nullified.

Also: This fixes handling of "manual jobs", which previously didn't handle jobs targeting more than `request_batch_size` due to not supporting updating `offset`

### CODE REVIEW

File: lib/sierra_manager.rb [hunk 1]
```diff
@@ -36,14 +36,19 @@ def fetch_updated_records
     while @processing
       threads = []
 
-      # Thread 1: Fetch next set of results:
+      # Thread 1: Encode previously fetcedFetch next set of results:
```
Comment (danamansana) [inline lib/sierra_manager.rb]: fetcedFetch?

File: lib/sierra_manager.rb [hunk 2]
```diff
@@ -55,10 +60,11 @@ def fetch_updated_records
   # If we have any previously retrieved Sierra response object waiting to be
   # sent to Kinesis, send it:
   def send_results_to_kinesis
-    unless @previous_results.nil?
-      sierra_batch = SierraBatch.new(@previous_results)
+    # @previous_results will be empty on the first run (before the first set of
+    # results have been received):
+    unless @previous_results.empty?
+      sierra_batch = SierraBatch.new(@previous_results.shift)
       sierra_batch.encode_and_send_to_kinesis if sierra_batch.has_results?
```
Comment (danamansana) [inline lib/sierra_manager.rb:68]: It looks like in the previous implementation we were waiting to see if the `sierra_batch.has_results?` before removing it from `@previous_results`, but in this implementation we `shift` it off the `@previous_results` right away. Should we be waiting? Does it make a difference?

**Review by danamansana — COMMENTED** [review]
Looks good, just a couple of small comments

File: lib/sierra_manager.rb (hunk 2 above)
Comment (nonword) [reply lib/sierra_manager.rb:68]: `sierra_batch.encode_and_send_to_kinesis` is still conditional on `sierra_batch.has_results?`. Previously `@previous_results` was a single resultset (or `nil`). It's now an array of resultsets (possibly empty). In both implementations, we only create a `SierraBatch` instance if `@previous_results` has data (i.e. not `nil` nor empty). And we then rely on `sierra_batch.has_results?` to determine whether or not to send them through kinesis. So I think the checks are all essentially the same. It's just the datatype of `@previous_results` that has changed.

**Pull request merged**